dvc/repo/reproduce.py

```diff
@@ -114,7 +119,7 @@ def reproduce(  # noqa: C901, PLR0912
         except RunCacheNotSupported as e:
             logger.warning("Failed to pull run cache: %s", e)
 
-    return _reproduce_stages(self.index.graph, list(stages), **kwargs)
+    return _reproduce_stages(self.index.graph, ldistinct(stages), **kwargs)
```

Comment:
Any reason for this change?

Reply:
I changed two things here: 
1. The order of `dvc repro --single-item stage1 stage2` always runs stage1 first and then stage2. Even if `stage2` is a dependency of `stage1`. Most users will just use a single target in `--single-item`. I did this for simplicity.
 Previously, it would run based on the order of the graph.
2. As `--single-item` targets are not processed, if you passed `dvc repro stage1 dvc.yaml`, 
 this would generate multiple stages. As I want to preserve the order for `--single-item`,  I replaced the `stages` to be a list instead of a set, but I still want to remove duplications (hence `ldistinct()`).

Reply:
Do we have tests for --single-item? (i couldn't find them)

> Most users will just use a single target in --single-item. I did this for simplicity.
Previously, it would run based on the order of the graph.

I didn't know that you could pass multiple targets to `single item`, unclear to me the product perspective cc @dberenbaum 

> As --single-item targets are not processed, if you passed dvc repro stage1 dvc.yaml,
this would generate multiple stages.

Sorry, I am not following the `dvc repro stage1 dvc.yaml`. What happens (before and after) if you put `--single-item` in that command

Reply:
> I didn't know that you could pass multiple targets to single item, unclear to me the product perspective cc @dberenbaum

I think that's okay. If you say `dvc repro -s stage1 stage2`, it'll run in that order.

> Sorry, I am not following the `dvc repro stage1 dvc.yaml`

`dvc.yaml` may contain multiple stages including `stage1` for example. So here we are just trying to deduplicate. The order of run will be, `stage1` followed by other stages present in `dvc.yaml` (except `stage1`).

Reply:
I don't think we need to clarify anything here except that `--single-item` runs only the given targets. Rest is implementation dependent/defined.

Reply:
Sorry to repeat, but are we testing all the behaviors you explained around `--single-item`? I just didn't find on a quick text search

Reply:
No test exists, because `—single-item` has been here from the early days.

https://github.com/iterative/dvc/blob/f6ded2fa72cd220ddebb46463ed6052d3dc286ab/dvc/command/repro.py#L37-L38

Reply:
Could we add some, since we are modifying the behavior in the P.R. might be a good time to do it, right?

Reply:
I thought `dvc repro --single-item stage1 stage2` isn't passing multiple targets to `--single-item` because `stage2` will be parsed as a positional arg target, right? Not sure whether we actually should be supporting positional targets and `--single-item` together, but no strong opinion because it seems like such a narrow use case.

Reply:
`--single-item` does not take any value, so `targets` is reused.
I think it makes sense for it to support multiple targets like `dvc repro -s stage1 stage2`. `--single-item` does not mean it takes one argument, it just means that we don't try to run it's dependencies or downward nodes and only run that target.

Reply:
> `--single-item` does not take any value, so `targets` is reused.

🤦 Okay, makes sense to me. If it's reasonable to support, I'd rather not have to make exceptions for when `repro` supports multiple vs single targets.

Reply:
FYI, another change here that we are discussing is that in `--single-item`, we won't run stages in order of topological order, but in order of the targets provided.

i.e.
```mermaid
graph TD;
      A-->B;
      A-->C;
```

If you do:
```console
dvc repro --single-item C A
```

The topological order of above graph is `[A, C]`. That's the order dvc runs right now. 
(In case of `--downstream`, it runs in `[C, A]` order).
But this PR changes it to `[C, A]`, in the order of targets. I don't think `--single-item` should care about the graph (we can even avoid collecting all stages just to run `dvc repro --single-item in the future).